### PR TITLE
Rework XAML compilation targets

### LINKED ISSUE #17099: InjectAvaloniaXamlOutput appears to mangle property IntermediateAssembly on multiple runs and prevents trimming
### Describe the bug

(For some reason this only appears on android builds..)

It looks like when InjectAvaloniaXamlOutput build target is run multiple times (for this case: (target "InjectAvaloniaXamlOutput" depends on it):)  which has already run because "Compile" target required it, the path becomes mangled.

```
CompileAvaloniaXamlInputs=
 816620                            obj/Release/net8.0-android/Avalonia/project.dll
 816621                                    AvaloniaCompileOutput=obj/Release/net8.0-android/Avalonia/Avalonia/project.dll
```

notice how there are now two /Avalonia/Avalonia nested directories when in fact all the build objects are only in Avalonia.

This in turn causes the trimmer targets to not look in the correct location for the assemblies which need to be trimmed

### To Reproduce

Compile android targets when trimming enabled.

### Expected behavior

previous version on which this worked without issue for me 11.1.999-cibuild0045786-beta. expect that it should also work in recent versions too.

### Avalonia version

11.1.2

### OS

Android

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by MrJul**
## What does the pull request do?
This PR reworks the MSBuild targets in `AvaloniaBuildTasks.targets`, responsible for compiling the XAML files into the final assembly.

## A bit of history
Before #13840, the original intermediate assembly (in the `obj` folder) was effectively patched in place (with a copy of the original one kept around).
This regularly caused some file access issues in parallel builds. It also made MSBuild think that the output was always changed between builds, effectively killing incremental builds.

After #13840, the original intermediate assembly is preserved, and the XAML compilation target outputs a new assembly instead. The intermediate assembly path is updated to the new one after the compilation.
At the time of that PR, this sounded like a good idea, preventing all the issues mentioned above.

However, this triggered a cascade of issues. Anything that relied on the intermediate assembly without triggering compilation was effectively broken, finding the old assembly instead.
This included NativeAOT, ClickOnce, publishing without building... Each subsequent fix triggering new failures in seemingly unrelated areas. For example, we've had #16835 re-trigger XAML compilation for `--no-build` publish, effectively breaking Android release builds which rely on publish targets implicitly, causing double compilation...

## How was the solution implemented?
What if instead we could have XAML compilation be part of the "standard" compilation (the `CoreCompile` target, responsible for running the C# compiler)?

That's possible with the [`TargetsTriggeredByCompilation`](https://github.com/dotnet/roslyn/blame/f6f0035c96393e68bb2e54b7118c38c20fdfb3b3/src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets#L181) property.
Any target referenced by this property will run as part of `CoreCompile`. We're free to modify the intermediate assembly here, since we didn't effectively "return" it from the target yet.

We also need to add the XAML files as inputs to that target, which is doable using the `CustomAdditionalCompileInputs` items.

With these changes, the target's logic is greatly simplified, and just naturally works, both for full rebuilds and incremental builds.

## Build tests

It was getting quite hard to ensure that we didn't regress anything while editing these targets.

To solve that, this PR introduces a new `BuildTests` solution, outside of the main one.
As part of CI, `BuildTests` is built using the NuGet packages that have just been generated from the main build. This ensures we're mimicking a standard user's solution as much as possible.

Output assemblies of the various build tests are then checked to ensure they correctly contain compiled XAML.
This is done for:
 - Standard C# platforms (Desktop, Android, iOS, Browser)
 - `publish --no-build`
 - NativeAOT
 - A WPF app containing both WPF and Avalonia XAML files
 - A F# project

## Fixed issues
 - Supersedes #13840
 - Supersedes #14966
 - Supersedes #15705
 - Supersedes #16835
 - Supersedes #17115
 - Supersedes #17145
 - Fixes #17099

### CODE REVIEW

File: nukebuild/Build.cs [hunk 1]
```diff
@@ -407,6 +408,67 @@ await Task.WhenAll(
...
+                    .SetConfiguration(configuration)
+                    .SetProperty("AvaloniaVersion", Parameters.Version)
+                    .SetProperty("NuGetPackageRoot", nugetCacheDirectory)
+                    .SetPackageDirectory(nugetCacheDirectory)
+                    .SetProjectFile(buildTestsDirectory / "BuildTests.sln")
+                    .SetProcessArgumentConfigurator(arguments => arguments.Add("--nodeReuse:false")));
+
+                // Standard compilation - should have compiled XAML
+                VerifyBuildTestAssembly("bin", "BuildTests");
+                VerifyBuildTestAssembly("bin", "BuildTests.Android");
```
Comment (kekekeks) [inline nukebuild/Build.cs:437]: Note that the original issue can't be reproduced with a plain `dotnet build`, with Rider it was only happened when one tries to debug the app.

**Review by maxkatz6 — APPROVED** [review]

**Comment by BobbyCannon** [comment]
11.2.2 does not work

11.3.999-cibuild0053376-alpha does work

**Comment by noorez** [comment]
It appears that even with that Alpha, the same issue that started this: https://github.com/AvaloniaUI/Avalonia/issues/17099 is happening again..  the obj path appears to be mangled again and the optimization step fails.

**Comment by MrJul** [comment]
@noorez 

Please check that your packages are all restored correctly, by verifying the effective versions in the solution explorer. This PR completely removes the extra assembly copy: the intermediate assembly path is never overwritten anymore, so you shouldn't see it at all.

We now even have a test checking this (the Android one, in release mode), which fails with a double Avalonia path before the changes, and passes after. This is also the case with your repository https://github.com/noorez/AvaloniaTest17099 which builds fine with version 11.3.999-cibuild0053376-alpha.

Please provide another MSBuild binary log is you encounter any further failure.

**Comment by noorez** [comment]
interestingly, even though I've explicitly used that nightly build version it appears that '11.3.999-cibuild0053380-alpha' got used instead. does this nightly contain that fix as well?

**Comment by MrJul** [comment]
> interestingly, even though I've explicitly used that nightly build version it appears that '11.3.999-cibuild0053380-alpha' got used instead. does this nightly contain that fix as well?

No, 53380 is a nightly build (from main branch) from 9 days ago. For PRs, you need the [nuget-feed-all](https://github.com/AvaloniaUI/Avalonia/wiki/Using-nightly-build-feed#all-builds-feed) feed set up.  That said, this PR is now merged so you should be able to use the very latest nightly build 53568.

**Comment by noorez** [comment]
ahh. I see what happened. I had that feed disabled by mistake :(. it builds as expected. 

apologies! Thanks!

**Pull request merged**